File: ios/runtime/XTRViewController.m
//
//  XTRViewController.m
//  XTSample
//
//  Created by 崔明辉 on 2017/8/25.
//  Copyright © 2017年 UED Center, YY Inc. All rights reserved.
//

#import "XTRViewController.h"
#import "XTRUtils.h"
#import "XTRContext.h"
#import "XTRNavigationController.h"
#import "XTRNavigationBar.h"
#import "XTRView.h"
#import <objc/runtime.h>
#import <XT-Mem/XTMemoryManager.h>

@interface XTRViewController ()

@property (nonatomic, assign) UIStatusBarStyle originalStatusBarStyle;
@property (nonatomic, strong) XTRNavigationBar *navigationBar;
@property (nonatomic, assign) BOOL navigationBarHidden;
@property (nonatomic, strong) UIView *innerView;
@property (nonatomic, weak) JSContext *context;

@end

@implementation XTRViewController

+ (NSString *)name {
    return @"XTRViewController";
}

+ (NSString *)create {
    XTRViewController *viewController = [XTRViewController new];
    XTManagedObject *managedObject = [[XTManagedObject alloc] initWithObject:viewController];
    [XTMemoryManager add:managedObject];
    viewController.context = [JSContext currentContext];
    viewController.objectUUID = managedObject.objectUUID;
    return managedObject.objectUUID;
}

#ifdef LOGDEALLOC
- (void)dealloc {
    NSLog(@"XTRViewController dealloc.");
}
#endif

- (instancetype)init
{
    self = [super init];
    if (self) {
        _originalStatusBarStyle = [UIApplication sharedApplication].statusBarStyle;
        _navigationBarHidden = YES;
    }
    return self;
}

- (JSValue *)scriptObject {
    return [self.context evaluateScript:[NSString stringWithFormat:@"objectRefs['%@']", self.objectUUID]];
}

+ (NSString *)xtr_view:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRViewController class]]) {
        XTRView *view = (id)[(XTRViewController *)obj innerView];
        if ([view isKindOfClass:[XTRView class]]) {
            return view.objectUUID;
        }
    }
    else if ([obj isKindOfClass:[UIViewController class]]) {
        XTRView *view = (id)obj.view;
        if ([view isKindOfClass:[XTRView class]]) {
            return view.objectUUID;
        }
    }
    return nil;
}

+ (void)xtr_setView:(NSString *)viewRef objectRef:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    UIView *view = [XTMemoryManager find:viewRef];
    if ([obj isKindOfClass:[UIViewController class]] && [view isKindOfClass:[UIView class]]) {
        obj.view = [UIView new];
        obj.view.backgroundColor = view.backgroundColor ?: [UIColor whiteColor];
        [obj.view addSubview:view];
        if ([obj isKindOfClass:[XTRViewController class]]) {
            [(XTRViewController *)obj setInnerView:view];
        }
    }
}

+ (NSDictionary *)xtr_safeAreaInsets:(NSString *)objectRef {
    XTRViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRViewController class]]) {
        CGFloat topLength = obj.topLayoutGuide.length;
        if (@available(iOS 11.0, *)) {
            topLength = obj.view.safeAreaInsets.top;
        }
        if (obj.navigationBar != nil && !obj.navigationBarHidden) {
            topLength += 44.0;
        }
        if (!obj.navigationBar.translucent) {
            topLength = 0;
        }
        return [JSValue fromInsets:UIEdgeInsetsMake(topLength, 0, 0, 0)];
    }
    return [JSValue fromInsets:UIEdgeInsetsZero];
}

+ (NSString *)xtr_parentViewController:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[UIViewController class]]) {
        id parentViewController = [obj parentViewController];
        if ([parentViewController conformsToProtocol:@protocol(XTRComponent)]) {
            return [parentViewController objectUUID];
        }
    }
    return nil;
}

+ (NSArray<NSString *> *)xtr_childViewControllers:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    NSMutableArray *childViewControllers = [NSMutableArray array];
    if ([obj isKindOfClass:[UIViewController class]]) {
        for (id viewController in obj.childViewControllers) {
            if ([viewController conformsToProtocol:@protocol(XTRComponent)]) {
                [childViewControllers addObject:[viewController objectUUID] ?: @""];
            }
        }
    }
    return [childViewControllers copy];
}

+ (void)xtr_addChildViewController:(NSString *)childControllerRef objectRef:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[UIViewController class]]) {
        UIViewController *childController = [XTMemoryManager find:childControllerRef];
        if ([childController isKindOfClass:[UIViewController class]]) {
            [obj addChildViewController:childController];
        }
    }
}

+ (void)xtr_removeFromParentViewController:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[UIViewController class]]) {
        [obj removeFromParentViewController];
    }
}

+ (NSString *)xtr_navigationController:(NSString *)objectRef {
    UIViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[UIViewController class]]) {
        UINavigationController *naviCtl = obj.navigationController;
        if ([naviCtl isKindOfClass:[XTRNavigationController class]]) {
            return [(XTRNavigationController *)naviCtl objectUUID];
        }
        else if ([naviCtl isKindOfClass:[UINavigationController class]]) {
            return [XTRNavigationController clone:naviCtl];
        }
    }
    return nil;
}

#pragma mark - ViewController callbacks

- (void)loadView {
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"loadView" withArguments:@[]];
        }
    }
}

- (void)viewDidLoad {
    [super viewDidLoad];
    self.automaticallyAdjustsScrollViewInsets = NO;
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewDidLoad" withArguments:@[]];
        }
    }
}

- (void)viewWillAppear:(BOOL)animated {
    [super viewWillAppear:animated];
    self.navigationItem.hidesBackButton = YES;
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewWillAppear" withArguments:@[]];
        }
    }
}

- (void)viewDidAppear:(BOOL)animated {
    [super viewDidAppear:animated];
    self.navigationItem.hidesBackButton = NO;
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewDidAppear" withArguments:@[]];
        }
    }
    if (self.navigationController != nil) {
        self.navigationController.navigationBar.hidden = YES;
    }
}

static UINavigationController *tmpNavigationController;
static BOOL onPanning;

- (void)viewWillDisappear:(BOOL)animated {
    [super viewWillDisappear:animated];
    self.navigationItem.hidesBackButton = YES;
    tmpNavigationController = self.navigationController;
    if (!onPanning) {
        [self viewWillForceDisappear:animated];
    }
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewWillDisappear" withArguments:@[]];
        }
    }
}

- (void)viewDidDisappear:(BOOL)animated {
    [super viewDidDisappear:animated];
    self.navigationItem.hidesBackButton = NO;
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewDidDisappear" withArguments:@[]];
        }
    }
    if (self.shouldRestoreNavigationBar && self.navigationController == nil) {
        tmpNavigationController.navigationBar.hidden = NO;
        tmpNavigationController.navigationBar.alpha = 1.0;
        [[UIApplication sharedApplication] setStatusBarStyle:self.originalStatusBarStyle];
    }
    tmpNavigationController = nil;
}

- (void)viewWillForceDisappear:(BOOL)animated {
    if ([self.navigationController.childViewControllers indexOfObject:self] == NSNotFound && self.shouldRestoreNavigationBar && animated) {
        tmpNavigationController.navigationBar.hidden = NO;
        [UIView animateWithDuration:0.25 animations:^{
            tmpNavigationController.navigationBar.alpha = 1.0;
        }];
    }
}

- (void)viewWillLayoutSubviews {
    [super viewWillLayoutSubviews];
    if (self.navigationBar != nil && !self.navigationBarHidden) {
        CGFloat topLength = self.topLayoutGuide.length + 44.0;
        if (@available(iOS 11.0, *)) {
            topLength = self.view.safeAreaInsets.top + 44.0;
        }
        self.navigationBar.frame = CGRectMake(0, 0, self.view.bounds.size.width, topLength);
        if (self.navigationBar.translucent) {
            self.innerView.frame = self.view.bounds;
        }
        else {
            self.innerView.frame = CGRectMake(0,
                                              topLength,
                                              self.view.bounds.size.width,
                                              self.view.bounds.size.height - topLength);
        }
    }
    else {
        self.innerView.frame = self.view.bounds;
    }
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewWillLayoutSubviews" withArguments:@[]];
        }
    }
}

- (void)viewDidLayoutSubviews {
    [super viewDidLayoutSubviews];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"viewDidLayoutSubviews" withArguments:@[]];
        }
    }
}

- (void)willMoveToParentViewController:(id)parent {
    [super willMoveToParentViewController:parent];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"_willMoveToParentViewController"
                  withArguments:[parent conformsToProtocol:@protocol(XTRComponent)] ? @[[parent objectUUID] ?: [NSNull null]] : @[]];
        }
    }
    if ([parent isKindOfClass:[UINavigationController class]] && self.shouldRestoreNavigationBar) {
        [self.navigationController.interactivePopGestureRecognizer
         addTarget:self
         action:@selector(onPopGesture:)];
    }
}

- (void)didMoveToParentViewController:(id)parent {
    [super didMoveToParentViewController:parent];
    if (self.scriptObject != nil) {
        JSValue *value = self.scriptObject;
        if (value != nil) {
            [value invokeMethod:@"_didMoveToParentViewController"
                  withArguments:[parent conformsToProtocol:@protocol(XTRComponent)] ? @[[parent objectUUID] ?: [NSNull null]] : @[]];
        }
    }
    if (parent == nil && self.exitAction) {
        self.exitAction(self);
    }
    if (parent == nil) {
        [tmpNavigationController.interactivePopGestureRecognizer
         removeTarget:self
         action:@selector(onPopGesture:)];
    }
}

- (void)onPopGesture:(UIPanGestureRecognizer *)sender {
    if (sender.state == UIGestureRecognizerStateBegan) {
        onPanning = YES;
    }
    else if (sender.state == UIGestureRecognizerStateEnded) {
        onPanning = NO;
    }
    CGFloat progress = [sender locationInView:self.view.window].x / self.view.window.bounds.size.width;
    if (![tmpNavigationController.childViewControllers containsObject:self] && self.shouldRestoreNavigationBar) {
        tmpNavigationController.navigationBar.hidden = NO;
        tmpNavigationController.navigationBar.alpha = progress;
    }
}

- (void)setNavigationBar:(XTRNavigationBar *)navigationBar {
    if (_navigationBar != nil) {
        [_navigationBar removeFromSuperview];
    }
    _navigationBar = navigationBar;
    if (_navigationBar != nil) {
        _navigationBar.viewController = self;
        [self.view addSubview:_navigationBar];
    }
    [self viewWillLayoutSubviews];
}

- (void)setNavigationBarLightContent:(BOOL)navigationBarLightContent {
    _navigationBarLightContent = navigationBarLightContent;
    [[UIApplication sharedApplication]
     setStatusBarStyle:_navigationBarLightContent ? UIStatusBarStyleLightContent : UIStatusBarStyleDefault];
    [self setNeedsStatusBarAppearanceUpdate];
}

- (UIStatusBarStyle)preferredStatusBarStyle {
    return self.navigationBarLightContent ? UIStatusBarStyleLightContent : UIStatusBarStyleDefault;
}

+ (NSString *)xtr_navigationBar:(NSString *)objectRef {
    XTRViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRViewController class]]) {
        return obj.navigationBar.objectUUID;
    }
    return nil;
}

+ (void)xtr_setNavigationBar:(NSString *)barRef objectRef:(NSString *)objectRef {
    XTRNavigationBar *bar = [XTMemoryManager find:barRef];
    XTRViewController *obj = [XTMemoryManager find:objectRef];
    if ([bar isKindOfClass:[XTRNavigationBar class]] && [obj isKindOfClass:[XTRViewController class]]) {
        obj.navigationBar = bar;
    }
}

+ (void)xtr_showNavigationBar:(BOOL)animated objectRef:(NSString *)objectRef {
    XTRViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRViewController class]]) {
        obj.navigationBarHidden = NO;
        obj.navigationBar.shouldShowBackBarButtonItem = [obj.navigationController.childViewControllers indexOfObject:obj] != 0;
        if (animated) {
            [UIView animateWithDuration:0.25 delay:0.0 usingSpringWithDamping:1.0 initialSpringVelocity:8.0 options:kNilOptions animations:^{
                obj.navigationBar.alpha = 1.0;
                [obj viewWillLayoutSubviews];
            } completion:nil];
        }
        else {
            obj.navigationBar.alpha = 1.0;
            [obj viewWillLayoutSubviews];
        }
    }
}

+ (void)xtr_hideNavigationBar:(BOOL)animated objectRef:(NSString *)objectRef {
    XTRViewController *obj = [XTMemoryManager find:objectRef];
    if ([obj isKindOfClass:[XTRViewController class]]) {
        obj.navigationBarHidden = YES;
        if (animated) {
            [UIView animateWithDuration:0.25 delay:0.0 usingSpringWithDamping:1.0 initialSpringVelocity:8.0 options:kNilOptions animations:^{
                obj.navigationBar.alpha = 0.0;
                [obj viewWillLayoutSubviews];
            } completion:nil];
        }
        else {
            obj.navigationBar.alpha = 0.0;
            [obj viewWillLayoutSubviews];
        }
    }
}

@end

File: android/xtruntime/src/main/java/com/opensource/xtruntime/XTRViewController.kt
package com.opensource.xtruntime

import android.app.Activity
import com.eclipsesource.v8.V8
import com.eclipsesource.v8.V8Array
import com.eclipsesource.v8.V8Object
import com.opensource.xtmem.XTManagedObject
import com.opensource.xtmem.XTMemoryManager
import com.opensource.xtruntime.libraries.keyboard.KeyboardHeightObserver
import com.opensource.xtruntime.libraries.keyboard.KeyboardHeightProvider
import java.lang.ref.WeakReference

/**
 * Created by cuiminghui on 2017/9/5.
 */
open class XTRViewController: XTRFragment(), XTRComponentInstance, KeyboardHeightObserver {

    override var objectUUID: String? = null
    lateinit var xtrContext: XTRContext

    override var view: XTRView? = null
        set(value) {
            if (field != null) { return }
            if (value == null) { return }
            field = value
            value.viewDelegate = WeakReference(this)
        }

    var parentViewController: WeakReference<XTRViewController>? = null
        internal set

    var childViewControllers: List<XTRViewController> = listOf()
        internal set

    fun setContentView(activity: Activity) {
        XTRScreen.resetScreenInfo(activity)
        this.requestFragment().let {
            val transaction = activity.fragmentManager.beginTransaction()
            transaction.replace(android.R.id.content, it)
            transaction.commit()
            it.setupKeyboardHeightProvider(activity)
        }
    }

    fun scriptObject(): V8Object? {
        return xtrContext.evaluateScript("objectRefs['$objectUUID']") as? V8Object
    }

    open fun requestFragment(): XTRViewController {
        return this
    }

    open fun viewDidLoad() {
        scriptObject()?.let {
            XTRContext.invokeMethod(it, "viewDidLoad")
            it.release()
        }
    }

    open fun viewWillAppear() {
        scriptObject()?.let {
            XTRContext.invokeMethod(it, "viewWillAppear")
            it.release()
        }
    }

    open fun viewDidAppear() {
        scriptObject()?.let {
            XTRContext.invokeMethod(it, "viewDidAppear")
            it.release()
        }
    }

    open fun viewWillDisappear() {
        scriptObject()?.let {
            XTRContext.invokeMethod(it, "viewWillDisappear")
            it.release()
        }
    }

    open fun viewDidDisappear() {
        scriptObject()?.let {
            XTRContext.invokeMethod(it, "viewDidDisappear")
            it.release()
        }
    }

    open fun viewWillLayoutSubviews() {
        scriptObject()?.let {
            XTRContext.invokeMethod(it, "viewWillLayoutSubviews")
            it.release()
        }
    }

    open fun viewDidLayoutSubviews() {
        scriptObject()?.let {
            XTRContext.invokeMethod(it, "viewDidLayoutSubviews")
            it.release()
        }
    }

    fun willMoveToParentViewController(parent: XTRViewController?) {
        scriptObject()?.let {
            XTRContext.invokeMethod(it, "_willMoveToParentViewController", listOf(
                    parent?.objectUUID ?: V8.getUndefined()
            ))
        }
    }

    fun didMoveToParentViewController(parent: XTRViewController?) {
        scriptObject()?.let {
            XTRContext.invokeMethod(it, "_didMoveToParentViewController", listOf(
                    parent?.objectUUID ?: V8.getUndefined()
            ))
        }
    }

    private var keyboardHeightProvider: KeyboardHeightProvider? = null

    private fun setupKeyboardHeightProvider(activity: Activity) {
        keyboardHeightProvider = KeyboardHeightProvider(activity)
        activity.findViewById(android.R.id.content).rootView?.post {
            keyboardHeightProvider?.start()
        }
    }

    override fun onDestroy() {
        super.onDestroy()
        keyboardHeightProvider?.close()
    }

    override fun onPause() {
        super.onPause()
        keyboardHeightProvider?.setKeyboardHeightObserver(null)
    }

    override fun onResume() {
        super.onResume()
        keyboardHeightProvider?.setKeyboardHeightObserver(this)
    }

    override fun onKeyboardHeightChanged(height: Int, orientation: Int) {
        this.lastKeyboardHeight = height
        handleKeyboardHeightChanged(height)
    }

    var lastKeyboardHeight = 0

    fun handleKeyboardHeightChanged(height: Int = this.lastKeyboardHeight) {
        scriptObject()?.let {
            if (height > 0) {
                XTRContext.invokeMethod(it, "keyboardWillShow", listOf(
                        XTRUtils.fromRect(XTRRect(
                                0.0,
                                0.0,
                                this.view?.bounds?.width ?: 0.0,
                                height.toDouble() / resources.displayMetrics.density), xtrContext.runtime), 0.25))
            }
            else {
                XTRContext.invokeMethod(it, "keyboardWillHide", listOf(0.25))
            }
            it.release()
        }
    }

    class JSExports(val context: XTRContext): XTRComponentExport() {

        override val name: String = "XTRViewController"

        override fun exports(): V8Object {
            val exports = V8Object(context.runtime)
            exports.registerJavaMethod(this, "create", "create", arrayOf())
            exports.registerJavaMethod(this, "xtr_view", "xtr_view", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_setView", "xtr_setView", arrayOf(String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_parentViewController", "xtr_parentViewController", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_childViewControllers", "xtr_childViewControllers", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_addChildViewController", "xtr_addChildViewController", arrayOf(String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_removeFromParentViewController", "xtr_removeFromParentViewController", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_navigationController", "xtr_navigationController", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_navigationBar", "xtr_navigationBar", arrayOf(String::class.java))
            exports.registerJavaMethod(this, "xtr_setNavigationBar", "xtr_setNavigationBar", arrayOf(String::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_showNavigationBar", "xtr_showNavigationBar", arrayOf(Boolean::class.java, String::class.java))
            exports.registerJavaMethod(this, "xtr_hideNavigationBar", "xtr_hideNavigationBar", arrayOf(Boolean::class.java, String::class.java))
            return exports
        }

        fun create(): String {
            val viewController = XTRViewController()
            viewController.xtrContext = context
            val managedObject = XTManagedObject(viewController)
            viewController.objectUUID = managedObject.objectUUID
            XTMemoryManager.add(managedObject)
            return managedObject.objectUUID
        }

        fun xtr_view(objectRef: String): String? {
            return (XTMemoryManager.find(objectRef) as? XTRViewController)?.view?.objectUUID
        }

        fun xtr_setView(viewRef: String, objectRef: String) {
            val view = XTMemoryManager.find(viewRef) as? XTRView ?: return
            val viewController = XTMemoryManager.find(objectRef) as? XTRViewController ?: return
            viewController.view = view
            view.post {
                viewController.viewDidLoad()
            }
        }

        fun xtr_parentViewController(objectRef: String): String? {
            return (XTMemoryManager.find(objectRef) as? XTRViewController)?.parentViewController?.get()?.objectUUID
        }

        fun xtr_childViewControllers(objectRef: String): V8Array? {
            return (XTMemoryManager.find(objectRef) as? XTRViewController)?.let {
                val v8Array = V8Array(context.runtime)
                it.childViewControllers.mapNotNull { it ->
                    return@mapNotNull it.objectUUID
                }.forEach { v8Array.push(it) }
                return@let v8Array
            }
        }

        fun xtr_addChildViewController(childControllerRef: String, objectRef: String) {
            val childController = XTMemoryManager.find(childControllerRef) as? XTRViewController ?: return
            (XTMemoryManager.find(objectRef) as? XTRViewController)?.let {
                if (childController.parentViewController == null) {
                    childController.willMoveToParentViewController(it)
                    it.childViewControllers.toMutableList()?.let { mutable ->
                        mutable.add(childController)
                        childController.parentViewController = WeakReference(it)
                        it.childViewControllers = mutable.toList()
                    }
                    childController.didMoveToParentViewController(it)
                }
            }
        }

        fun xtr_removeFromParentViewController(objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTRViewController)?.let {
                it.parentViewController?.let { parentViewController ->
                    it.willMoveToParentViewController(null)
                    parentViewController.get()?.childViewControllers?.toMutableList()?.let { mutable ->
                        mutable.remove(it)
                        parentViewController.get()?.childViewControllers = mutable.toList()
                    }
                    it.didMoveToParentViewController(null)
                }
                it.parentViewController = null
            }
        }

        fun xtr_navigationController(objectRef: String): String? {
            (XTMemoryManager.find(objectRef) as? XTRViewController)?.let {
                var currentParentViewController = it.parentViewController?.get()
                while (currentParentViewController != null) {
                    if (currentParentViewController is XTRNavigationController) {
                        return currentParentViewController.objectUUID
                    }
                    currentParentViewController = currentParentViewController.parentViewController?.get()
                }
            }
            return null
        }

        fun xtr_navigationBar(objectRef: String): String? {
            return (XTMemoryManager.find(objectRef) as? XTRViewController)?.requestFragment()?.navigationBar?.objectUUID
        }

        fun xtr_setNavigationBar(navigationBarRef: String, objectRef: String) {
            val navigationBar = XTMemoryManager.find(navigationBarRef) as? XTRNavigationBar ?: return
            (XTMemoryManager.find(objectRef) as? XTRViewController)?.requestFragment()?.navigationBar = navigationBar
        }

        fun xtr_showNavigationBar(value: Boolean, objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTRViewController)?.requestFragment()?.navigationBarHidden = false
        }

        fun xtr_hideNavigationBar(value: Boolean, objectRef: String) {
            (XTMemoryManager.find(objectRef) as? XTRViewController)?.requestFragment()?.navigationBarHidden = true
        }

    }

}